### DISCUSSION TITLE
How to estimate water (or any polygon area or line length) within each h3 index

### DISCUSSION THREAD

**Discussion by tonialmagro-tomtom** [uber/h3#910] [General]
H3 index are great to work with points, however I would like to know an efficient way to given a big polygon in the map (imaging a lake or a river), how can we estimate the area of water within each h3 index in the map?

**Comment by nrabinowitz** [comment]
There are various approaches here, but the grid-centric one, which is likely the most efficient, likely looks something like this:

 1. Given target H3 cells at res `r`, use `polygonToCells` to convert the water feature to a set of H3 cells at res `r+o`. The greater the value of offset `o`, the more processing and memory you'll need for the operation (approximately 7x increase per 1 resolution), but the more precision you'll get in the output. I'd probably go with `r+2` here.
 2. For each target cell, use `cellToChildren(cell, r+o)` to get the children at the water resolution. Count the number of these children that are in the water set. The value of `children in water / children` is the percentage of the cell covered by water. If you need the area, you could sum the child areas using `cellAreaKm`, or just use `cellAreaKm` on the parent and take the percentage.

Notes:
- This is only an estimate, and there are various sources of potential error. One good reason to not use offset `o > 2` is that some of the water cells might actually be [fully outside the area of the parent](https://observablehq.com/@nrabinowitz/h3-hierarchical-non-containment).
- If the area covered by the target cells is expected to be much larger than the water area, you might turn the algorithm around and group the cells by `cellToParent(waterCell, r)`, resulting in a hashmap of parent cell to water cell count, then look this number up for every target cell. This might actually be faster in both cases, since creating the hashmap might be as cheap as creating the set (both are `O(n)`), and the hashmap approach avoids calling `cellToChildren` on each target cell.
